# Wrangling, Analyzing, and Visualizing WeRateDogs Tweets

## Gather
1. We have given the data in a CSV file `twitter-archive-enhanced.csv`
<br></br>
2. We should gather additional data via `Twitter API` for (retweets count and likes count)
<br></br>
3. Image predictions it is a file hosted in a webpage. We need to daownload it programmatically using `Requests` library

In [1]:
import pandas as pd
import json
import requests
import io

In [2]:
archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

In [4]:
urlData = requests.get(url).content
image_prediction = pd.read_csv(io.StringIO(urlData.decode('utf-8')), sep='\t')

In [5]:
tweet_details = pd.DataFrame(columns=['tweet_id', 'retweet_count', 'favorite_count'])
with open('tweet-json.txt') as f:
    for line in f:
        status  = json.loads(line)
        tweet_id = status['id_str']
        retweet_count = status['retweet_count']
        favorite_count = status['favorite_count']
        tweet_details = tweet_details .append(pd.DataFrame([[tweet_id, retweet_count, favorite_count]],
                                        columns=['tweet_id', 'retweet_count', 'favorite_count']))
tweet_details = tweet_details .reset_index(drop=True)

### Twitter API
Does not work

## Assess

In [6]:
# Let's assess archive table visually
archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [7]:
# Let's assess archive table programmatically 
archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [8]:
# What is the column in_reply_to_status_is for
archive[archive['in_reply_to_status_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
149   863079547188785154           6.671522e+17         4.196984e+09   
179   857214891891077121           8.571567e+17         1.806710e+08   
184   856526610513747968           8.558181e+17         4.196984e+09   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
251   844979544864018432           7.590995e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
565   802265048156610565           7.331095e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
...                  ...                    ...                  ...   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1501  692142790915014657           6.920419e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1630  684538444857667585           6.844811e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1842  675870721063669760           6.757073e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1852  675707330206547968           6.754971e+17         4.196984e+09   
1866  675349384339542016           6.749998e+17         4.196984e+09   
1882  674999807681908736           6.747934e+17         4.196984e+09   
1885  674793399141146624           6.717299e+17         4.196984e+09   
1892  674754018082705410           6.747522e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [9]:
# What is the column retweeted_status_id for
archive[archive['retweeted_status_id'].notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
78    877611172832227328                    NaN                  NaN   
91    874434818259525634                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
109   871166179821445120                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
130   867072653475098625                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
137   866094527597207552                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
160   860924035999428608                    NaN                  NaN   
165   860177593139703809                    NaN                  NaN   
171   858860390427611136                    NaN                  NaN   
180   857062103051644929                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
185   856330835276025856                    NaN                  NaN   
194   855245323840757760                    NaN                  NaN   
195   855138241867124737                    NaN                  NaN   
204   852936405516943360                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
...                  ...                    ...                  ...   
784   775096608509886464                    NaN                  NaN   
794   773336787167145985                    NaN                  NaN   
800   772615324260794368                    NaN                  NaN   
811   771171053431250945                    NaN                  NaN   
815   771004394259247104                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
826   769335591808995329                    NaN                  NaN   
829   768909767477751808                    NaN                  NaN   
833   768554158521745409                    NaN                  NaN   
841   766864461642756096                    NaN                  NaN   
847   766078092750233600                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
868   761750502866649088                    NaN                  NaN   
872   761371037149827077                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
890   759566828574212096                    NaN                  NaN   
895   759159934323924993                    NaN                  NaN   
908   757729163776290825                    NaN                  NaN   
911   757597904299253760                    NaN                  NaN   
926   754874841593970688                    NaN                  NaN   
937   753298634498793472                    NaN                  NaN   
943   752701944171524096                    NaN                  NaN   
949   752309394570878976                    NaN         

In [10]:
# Check for duplicate in tweet_id
archive[archive.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [11]:
# Check for duplicate in expanded_urls
archive[archive.expanded_urls.duplicated()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
75    878281511006478336                    NaN                  NaN   
76    878057613040115712                    NaN                  NaN   
98    873213775632977920                    NaN                  NaN   
113   870726314365509632           8.707262e+17         1.648776e+07   
126   868552278524837888                    NaN                  NaN   
135   866450705531457537                    NaN                  NaN   
136   866334964761202691                    NaN                  NaN   
148   863427515083354112           8.634256e+17         7.759620e+07   
150   863062471531167744                    NaN                  NaN   
161   860563773140209665                    NaN                  NaN   
169   859196978902773760                    NaN                  NaN   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
253   844704788403113984                    NaN                  NaN   
264   842163532590374912                    NaN                  NaN   
271   841077006473256960                    NaN                  NaN   
274   840698636975636481           8.406983e+17         8.405479e+17   
276   840632337062862849                    NaN                  NaN   
280   839549326359670784                    NaN                  NaN   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
293   837820167694528512                    NaN                  NaN   
...                  ...                    ...                  ...   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1655  683391852557561860                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1683  681694085539872773                    NaN                  NaN   
1688  681523177663676416                    NaN                  NaN   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1726  680055455951884288                    NaN                  NaN   
1742  679462823135686656                    NaN                  NaN   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1806  676936541936185344                    NaN                  NaN   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         4.196984e+09   
1970  673295268553605120                    NaN                  NaN   
2038  671550332464455680           6.715449e+17         4.196984e+09   
2149  669684865554620416           6.693544e+17         4.196984e+09   
2180  669000397445533696                    NaN                  NaN   
2189  668967877119254528           6.689207e+17         2.143566e+07   
2212  668587383441514497                    NaN         

In [61]:
archive.query('expanded_urls == "https://www.gofundme.com/3yd6y1c,https://twitter.com/dog_rates/status/878281511006478336/photo/1"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
73  878404777348136964                    NaN                  NaN   
75  878281511006478336                    NaN                  NaN   

                    timestamp  \
73  2017-06-24 00:09:53 +0000   
75  2017-06-23 16:00:04 +0000   

                                               source  \
73  <a href="http://twitter.com/download/iphone" r...   
75  <a href="http://twitter.com/download/iphone" r...   

                                                 text  retweeted_status_id  \
73  RT @dog_rates: Meet Shadow. In an attempt to r...         8.782815e+17   
75  Meet Shadow. In an attempt to reach maximum zo...                  NaN   

    retweeted_status_user_id retweeted_status_timestamp  \
73              4.196984e+09  2017-06-23 16:00:04 +0000   
75                       NaN                        NaN   

                                        expanded_urls  rating_numerator  \
73  https://www.gofundme.com/3yd6y1c,https://twitt...                13   
75  https://www.gofundme.com/3yd6y1c,https://twitt...                13   

    rating_denominator    name doggo floofer pupper puppo  
73                  10  Shadow  None    None   None  None  
75                  10  Shadow  None    None   None  None

In [62]:
archive.query('expanded_urls == "https://vine.co/v/ea0OwvPTx9l"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
657   791774931465953280                    NaN                  NaN   
2212  668587383441514497                    NaN                  NaN   

                      timestamp  \
657   2016-10-27 22:53:48 +0000   
2212  2015-11-23 00:30:28 +0000   

                                                 source  \
657   <a href="http://vine.co" rel="nofollow">Vine -...   
2212  <a href="http://vine.co" rel="nofollow">Vine -...   

                                                   text  retweeted_status_id  \
657   Vine will be deeply missed. This was by far my...                  NaN   
2212  Never forget this vine. You will not stop watc...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
657                        NaN                        NaN   
2212                       NaN                        NaN   

                      expanded_urls  rating_numerator  rating_denominator  \
657   https://vine.co/v/ea0OwvPTx9l                14                  10   
2212  https://vine.co/v/ea0OwvPTx9l                13                  10   

      name doggo floofer pupper puppo  
657   None  None    None   None  None  
2212   the  None    None   None  None

In [12]:
# What is the not null value(s) of column doggo
archive[archive['doggo']!= 'None']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
9     890240255349198849                    NaN                  NaN   
43    884162670584377345                    NaN                  NaN   
99    872967104147763200                    NaN                  NaN   
108   871515927908634625                    NaN                  NaN   
110   871102520638267392                    NaN                  NaN   
121   869596645499047938                    NaN                  NaN   
172   858843525470990336                    NaN                  NaN   
191   855851453814013952                    NaN                  NaN   
200   854010172552949760                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
240   846514051647705089                    NaN                  NaN   
248   845397057150107648                    NaN                  NaN   
300   836753516572119041                    NaN                  NaN   
318   834574053763584002                    NaN                  NaN   
323   834089966724603904                    NaN                  NaN   
331   832998151111966721                    NaN                  NaN   
339   832273440279240704                    NaN                  NaN   
344   832032802820481025                    NaN                  NaN   
345   831939777352105988                    NaN                  NaN   
351   831322785565769729                    NaN                  NaN   
359   829878982036299777                    NaN                  NaN   
362   829449946868879360                    NaN                  NaN   
363   829374341691346946                    NaN                  NaN   
372   828381636999917570                    NaN                  NaN   
384   826958653328592898                    NaN                  NaN   
385   826848821049180160                    NaN                  NaN   
389   826476773533745153                    NaN                  NaN   
391   826204788643753985                    NaN                  NaN   
423   821765923262631936                    NaN                  NaN   
425   821421320206483457                    NaN                  NaN   
...                  ...                    ...                  ...   
857   763956972077010945           7.638652e+17         1.584641e+07   
877   760893934457552897                    NaN                  NaN   
881   760521673607086080                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
899   758828659922702336                    NaN                  NaN   
914   757393109802180609                    NaN                  NaN   
919   756526248105566208                    NaN                  NaN   
924   755206590534418437                    NaN                  NaN   
944   752682090207055872                    NaN                  NaN   
945   752660715232722944                    NaN                  NaN   
948   752334515931054080                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
967   750381685133418496           7.501805e+17         4.717297e+09   
977   750011400160841729                    NaN                  NaN   
985   749075273010798592                    NaN                  NaN   
989   748932637671223296                    NaN                  NaN   
992   748692773788876800                    NaN                  NaN   
1030  745433870967832576                    NaN                  NaN   
1039  744234799360020481                    NaN                  NaN   
1051  742534281772302336                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1075  739623569819336705                    NaN                  NaN   
1079  739238157791694849                    NaN                  NaN   
1103  735256018284875776                    NaN         

In [13]:
# What is the not null value(s) of column doggo
archive.doggo.unique()

array(['None', 'doggo'], dtype=object)

In [14]:
# How many values in column rating_numerator
archive.rating_numerator.unique()

array([  13,   12,   14,    5,   17,   11,   10,  420,  666,    6,   15,
        182,  960,    0,   75,    7,   84,    9,   24,    8,    1,   27,
          3,    4,  165, 1776,  204,   50,   99,   80,   45,   60,   44,
        143,  121,   20,   26,    2,  144,   88])

In [15]:
# what does the rating_denominator have if rating_denominator = 170
archive[archive['rating_denominator'] == 170]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1120  731156023742988288                    NaN                  NaN   

                      timestamp  \
1120  2016-05-13 16:15:54 +0000   

                                                 source  \
1120  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1120  Say hello to this unbelievably well behaved sq...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1120                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1120  https://twitter.com/dog_rates/status/731156023...               204   

      rating_denominator  name doggo floofer pupper puppo  
1120                 170  this  None    None   None  None

In [16]:
# How many different value are there in rating_numerator column
archive['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [17]:
# How many different values in rating_denominator column
archive.rating_denominator.unique()

array([ 10,   0,  15,  70,   7,  11, 150, 170,  20,  50,  90,  80,  40,
       130, 110,  16, 120,   2])

In [18]:
# Let's count the number of different values
archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [19]:
archive.name.value_counts()

None         745
a             55
Charlie       12
Oliver        11
Cooper        11
Lucy          11
Tucker        10
Penny         10
Lola          10
Winston        9
Bo             9
Sadie          8
the            8
Bailey         7
an             7
Daisy          7
Toby           7
Buddy          7
Milo           6
Dave           6
Oscar          6
Jax            6
Bella          6
Jack           6
Rusty          6
Stanley        6
Koda           6
Leo            6
Scout          6
Sunny          5
            ... 
Genevieve      1
Zuzu           1
Shikha         1
Tanner         1
Linus          1
Creg           1
Ralph          1
Jockson        1
River          1
Brownie        1
Antony         1
Halo           1
Toffee         1
Biden          1
Obie           1
Bruiser        1
Perry          1
Samsom         1
Joey           1
Beebop         1
Lucia          1
Dug            1
Vixen          1
Orion          1
Tuck           1
Jaycob         1
Carbon         1
all           

In [20]:
# How many rows having one stage "doggo or floofer or pupper or puppo"
archive.query('doggo == "doggo" or floofer =="floofer" or pupper == "pupper" or puppo == "puppo"').tweet_id.count()

380

In [21]:
archive.name.value_counts()

None         745
a             55
Charlie       12
Oliver        11
Cooper        11
Lucy          11
Tucker        10
Penny         10
Lola          10
Winston        9
Bo             9
Sadie          8
the            8
Bailey         7
an             7
Daisy          7
Toby           7
Buddy          7
Milo           6
Dave           6
Oscar          6
Jax            6
Bella          6
Jack           6
Rusty          6
Stanley        6
Koda           6
Leo            6
Scout          6
Sunny          5
            ... 
Genevieve      1
Zuzu           1
Shikha         1
Tanner         1
Linus          1
Creg           1
Ralph          1
Jockson        1
River          1
Brownie        1
Antony         1
Halo           1
Toffee         1
Biden          1
Obie           1
Bruiser        1
Perry          1
Samsom         1
Joey           1
Beebop         1
Lucia          1
Dug            1
Vixen          1
Orion          1
Tuck           1
Jaycob         1
Carbon         1
all           

In [22]:
# View the name of columns to describe what each one of them contains
list(archive)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

`archive` columns:
- **'tweet_id'**: the unique identifier for each tweet. And it is the last part in the tweet's url
- **'in_reply_to_status_id'**: Nullable. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s ID
- **'in_reply_to_user_id'**: Nullable. If the represented Tweet is a reply, this field will contain the integer representation of the original Tweet’s author ID. This will not necessarily always be the user directly mentioned in the Tweet.
- **'timestamp'**: This will contain the tweet date & time
- **'source'**: Utility used to post the Tweet, as an HTML-formatted string. Tweets from the Twitter website have a source value of web.
- **'text'**: The actual UTF-8 text of the status update. See twitter-text for details on what characters are currently considered valid.
- **'retweeted_status_id'**: This will be not null if the tweet is a retweeted tweet, and it will contain the tweet's id
- **'retweeted_status_user_id'**: This will be not null if the tweet is a retweeted tweet, and it will contain the tweet's user id
- **'retweeted_status_timestamp'**: This will be not null if the tweet is a retweeted tweet, and it will contain the tweet's date & time
- **'expanded_urls'**: The full link for the tweet
- **'rating_numerator'**: Rating value Numerator will be one of these values [13, 12, 14, 5, 17, 11, 10, 420, 666, 6, 15, 182, 960, 0, 75, 7, 84, 9, 24, 8, 1, 27, 3, 4, 165, 1776, 204, 50, 99, 80, 45, 60, 44, 143, 121, 20, 26, 2, 144, 88]
- **'rating_denominator'**: Rating value Denominator will be one of these values [10, 0,15, 70, 7, 11, 150, 170, 20, 50, 90, 80, 40, 130, 110, 16, 120, 2]
- **'name'**: Dog Name
- **'doggo'**: Dog Stage, and its value either ['None', 'doggo']. Note: None is a text not NaN.
- **'floofer'**: Dog Stage, and its value either ['None', 'floofer']. Note: None is a text not NaN.
- **'pupper'**: Dog Stage, and its value either ['None', 'pupper']. Note: None is a text not NaN.
- **'puppo'**: Dog Stage, and its value either ['None', 'puppo']. Note: None is a text not NaN.

In [23]:
# Let's assess tweet_details table visually
tweet_details

tweet_id retweet_count favorite_count
0     892420643555336193          8853          39467
1     892177421306343426          6514          33819
2     891815181378084864          4328          25461
3     891689557279858688          8964          42908
4     891327558926688256          9774          41048
5     891087950875897856          3261          20562
6     890971913173991426          2158          12041
7     890729181411237888         16716          56848
8     890609185150312448          4429          28226
9     890240255349198849          7711          32467
10    890006608113172480          7624          31166
11    889880896479866881          5156          28268
12    889665388333682689          8538          38818
13    889638837579907072          4735          27672
14    889531135344209921          2321          15359
15    889278841981685760          5637          25652
16    888917238123831296          4709          29611
17    888804989199671297          4559          26080
18    888554962724278272          3732          20290
19    888078434458587136          3653          22201
20    887705289381826560          5609          30779
21    887517139158093824         12082          46959
22    887473957103951883         18781          69871
23    887343217045368832         10737          34222
24    887101392804085760          6167          31061
25    886983233522544640          8084          35859
26    886736880519319552          3443          12306
27    886680336477933568          4610          22798
28    886366144734445568          3316          21524
29    886267009285017600             4            117
...                  ...           ...            ...
2324  666411507551481857           339            459
2325  666407126856765440            44            113
2326  666396247373291520            92            172
2327  666373753744588802           100            194
2328  666362758909284353           595            804
2329  666353288456101888            77            229
2330  666345417576210432           146            307
2331  666337882303524864            96            204
2332  666293911632134144           368            522
2333  666287406224695296            71            152
2334  666273097616637952            82            184
2335  666268910803644416            37            108
2336  666104133288665088          6871          14765
2337  666102155909144576            16             81
2338  666099513787052032            73            164
2339  666094000022159362            79            169
2340  666082916733198337            47            121
2341  666073100786774016           174            335
2342  666071193221509120            67            154
2343  666063827256086533           232            496
2344  666058600524156928            61            115
2345  666057090499244032           146            304
2346  666055525042405380           261            448
2347  666051853826850816           879           1253
2348  666050758794694657            60            136
2349  666049248165822465            41            111
2350  666044226329800704           147            311
2351  666033412701032449            47            128
2352  666029285002620928            48            132
2353  666020888022790149           532           2535

[2354 rows x 3 columns]

In [24]:
# Let's assess archive table programmically
tweet_details.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null object
retweet_count     2354 non-null object
favorite_count    2354 non-null object
dtypes: object(3)
memory usage: 55.2+ KB


In [25]:
tweet_details[tweet_details.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, retweet_count, favorite_count]
Index: []

In [26]:
# describe function does not work because of the type of columns
tweet_details.describe()

tweet_id  retweet_count  favorite_count
count                 2354           2354            2354
unique                2354           1724            2007
top     707038192327901184           3652               0
freq                     1              5             179

In [27]:
# Let's assess image_prediction table visually
image_prediction

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [28]:
image_prediction[image_prediction['p1'].str.match('s')].p1.value_counts()

seat_belt                      22
soft-coated_wheaten_terrier    11
schipperke                     10
standard_poodle                 8
swing                           7
shopping_cart                   5
stone_wall                      3
seashore                        3
space_heater                    3
sea_lion                        3
ski_mask                        3
sorrel                          2
shower_curtain                  2
snail                           2
snorkel                         2
street_sign                     2
school_bus                      1
sliding_door                    1
shopping_basket                 1
skunk                           1
shield                          1
starfish                        1
slug                            1
sunglasses                      1
sea_urchin                      1
silky_terrier                   1
sulphur-crested_cockatoo        1
stove                           1
studio_couch                    1
scorpion      

In [29]:
image_prediction[image_prediction['p1'].str.match('S')].p1.value_counts()

Samoyed                      43
Staffordshire_bullterrier    20
Siberian_husky               20
Shetland_sheepdog            18
Shih-Tzu                     17
Siamese_cat                   7
Saint_Bernard                 7
Saluki                        4
Scottish_deerhound            3
Sussex_spaniel                2
Scotch_terrier                1
Name: p1, dtype: int64

In [30]:
# is there any image has the same breed prediction with different values of p1, p2, and p3
image_prediction.query('p1 == p2 or p1 == p3 or p2 == p3')

tweet_id                                          jpg_url  \
1409  770414278348247044  https://pbs.twimg.com/media/CrEPsfWXEAAKvem.jpg   

      img_num       p1   p1_conf  p1_dog       p2   p2_conf  p2_dog  \
1409        1  maillot  0.580528   False  maillot  0.081449   False   

                    p3  p3_conf  p3_dog  
1409  golden_retriever  0.05357    True

In [31]:
# is p1_conf the most prediction value 
image_prediction.query('p1_conf > p2_conf and p1_conf > p3_conf').count()[0]

2075

In [32]:
image_prediction.query('img_num == 4')

tweet_id                                          jpg_url  \
144   668623201287675904  https://pbs.twimg.com/media/CUdtP1xUYAIeBnE.jpg   
779   689905486972461056  https://pbs.twimg.com/media/CZMJYCRVAAE35Wk.jpg   
1024  710588934686908417  https://pbs.twimg.com/media/CdyE2x1W8AAe0TG.jpg   
1161  734787690684657664  https://pbs.twimg.com/media/CjJ9gQ1WgAAXQtJ.jpg   
1286  750868782890057730  https://pbs.twimg.com/media/CmufLLsXYAAsU0r.jpg   
1325  756998049151549440  https://pbs.twimg.com/media/CoFlsGAWgAA2YeV.jpg   
1337  758405701903519748  https://pbs.twimg.com/media/CoZl9fXWgAMox0n.jpg   
1342  758854675097526272  https://pbs.twimg.com/media/Cof-SuqVYAAs4kZ.jpg   
1372  762464539388485633  https://pbs.twimg.com/media/CpTRc4DUEAAYTq6.jpg   
1437  773985732834758656  https://pbs.twimg.com/media/Cr2_6R8WAAAUMtc.jpg   
1496  783391753726550016  https://pbs.twimg.com/media/Ct8qn8EWIAAk9zP.jpg   
1542  791312159183634433  https://pbs.twimg.com/media/CvtONV4WAAAQ3Rn.jpg   
1543  791406955684368384  https://pbs.twimg.com/media/CvukbEkWAAAV-69.jpg   
1549  792913359805018113  https://pbs.twimg.com/media/CwD-eCLWIAA6v0B.jpg   
1617  802323869084381190  https://pbs.twimg.com/media/CyJtSmDUAAA2F9x.jpg   
1713  818614493328580609  https://pbs.twimg.com/media/C1xNgraVIAA3EVb.jpg   
1716  819006400881917954  https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1717  819015331746349057  https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1766  826848821049180160  https://pbs.twimg.com/media/C3mOnZ_XUAAjr2V.jpg   
1768  827199976799354881  https://pbs.twimg.com/media/C3rN-lcWEAA9CmR.jpg   
1790  830097400375152640  https://pbs.twimg.com/media/C4UZLZLWYAA0dcs.jpg   
1795  831315979191906304  https://pbs.twimg.com/media/C4lst0bXAAE6MP8.jpg   
1800  831911600680497154  https://pbs.twimg.com/media/C4uLLGuUoAAkIHm.jpg   
1829  836001077879255040  https://pbs.twimg.com/media/C5oSiskU0AE8sJ_.jpg   
1893  849412302885593088  https://pbs.twimg.com/media/C8m3-iQVoAAETnF.jpg   
1914  854120357044912130  https://pbs.twimg.com/media/C9px7jyVwAAnmwN.jpg   
1920  856282028240666624  https://pbs.twimg.com/media/C-If9ZwXoAAfDX2.jpg   
1934  859851578198683649  https://pbs.twimg.com/media/C-7OcfyXsAAsqzU.jpg   
1954  864197398364647424  https://pbs.twimg.com/media/C_4-8iPV0AA1Twg.jpg   
1978  870656317836468226  https://pbs.twimg.com/media/DBUxSSTXsAA-Jn1.jpg   
2040  885167619883638784  https://pbs.twimg.com/media/DEi_N9qXYAAgEEw.jpg   

      img_num                        p1   p1_conf  p1_dog  \
144         4                 Chihuahua  0.708163    True   
779         4                Pomeranian  0.943331    True   
1024        4                  Pembroke  0.982004    True   
1161        4          golden_retriever  0.883991    True   
1286        4                toy_poodle  0.912648    True   
1325        4          golden_retriever  0.678555    True   
1337        4  Chesapeake_Bay_retriever  0.702954    True   
1342        4                    barrow  0.974047   False   
1372        4                      chow  0.999953    True   
1437        4               giant_panda  0.451149   False   
1496        4        Norwegian_elkhound  0.877130    True   
1542        4        miniature_pinscher  0.892925    True   
1543        4                  Pembroke  0.972629    True   
1549        4                  web_site  0.226716   False   
1617        4              home_theater  0.765069   False   
1713        4                 Chihuahua  0.450722    True   
1716        4                    prison  0.907083   False   
1717        4                    prison  0.907083   False   
1766        4            Great_Pyrenees  0.858764    True   
1768        4                Great_Dane  0.869681    True   
1790        4                toy_poodle  0.442713    True   
1795        4                    briard  0.982755    True   
1800        4                bloodhound  0.777562    True   
1829        4                   Samoyed  0.963558    True   
1893        4             

In [33]:
# Let's assess image_prediction table programmatically 
image_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [34]:
# View the name of columns to describe what each one of them contains
list(image_prediction)

['tweet_id',
 'jpg_url',
 'img_num',
 'p1',
 'p1_conf',
 'p1_dog',
 'p2',
 'p2_conf',
 'p2_dog',
 'p3',
 'p3_conf',
 'p3_dog']

`image_prediction` columns:
<br></br>
- **tweet_id**: is the last part of the tweet URL after "status/"
- **jpg_url**: is the link for the image used in the algorithms
- **img_num**: is the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images)
- **p1**: is the algorithm's #1 prediction for the image in the tweet
- **p1_conf**: is how confident the algorithm is in its #1 prediction
- **p1_dog**: is whether or not the #1 prediction is a breed of dog
- **p2**: is the algorithm's second most likely prediction
- **p2_conf**: is how confident the algorithm is in its #2 prediction
- **p2_dog**: is whether or not the #2 prediction is a breed of dog
- **p3**: is the algorithm's third most likely prediction
- **p3_conf**: is how confident the algorithm is in its #3 prediction
- **p3_dog**: is whether or not the #3 prediction is a breed of dog

In [35]:
# Let's see if there are duplicated values based on jpg_url
image_prediction[image_prediction.jpg_url.duplicated()]

tweet_id                                            jpg_url  \
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1315  754874841593970688    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333  757729163776290825    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349  759566828574212096    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
1364  761371037149827077  https://pbs.twimg.com/tweet_video_thumb/CeBym7...   
1368  761750502866649088    https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   
1387  766078092750233600    https://pbs.twimg.com/media/ChK1tdBWwAQ1flD.jpg   
1407  770093767776997377    https://pbs.twimg.com/media/CkjMx99UoAM2B1a.jpg   
1417  771171053431250945    https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg   
1427  772615324260794368    https://pbs.twimg.com/media/Cp6db4-XYAAMmqL.jpg   
1446  775898661951791106    https://pbs.twimg.com/media/CiyHLocU4AI2pJu.jpg   
1453  776819012571455488    https://pbs.twimg.com/media/CW88XN4WsAAlo8r.jpg   
1456  777641927919427584    https://pbs.twimg.com/media/CmoPdmHW8AAi8BI.jpg   
1463  778396591732486144    https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg   
1476  780496263422808064    https://pbs.twimg.com/media/Ck2d7tJWUAEPTL3.jpg   
1487  782021823840026624    https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
1495  783347506784731136    https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg   
1510  786036967502913536    https://pbs.twimg.com/media/CtKHLuCWYAA2TTs.jpg   
1522  788070120937619456    https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg   
1538  790723298204217344    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1541  791026214425268224    https://pbs.twimg.com/media/CpmyNumW8AAAJGj.jpg   
1564  793614319594401792    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1569  794355576146903043    https://pbs.twimg.com/media/CvJCabcWgAIoUxW.jpg   
1571  794983741416415232    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg   
1579  796177847564038144    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg   
1588  798340744599797760    https://pbs.twimg.com/media/CrXhIqBW8AA6Bse.jpg   
1589  798628517273620480    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
1590  798644042770751489    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
1591  798665375516884993    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
...                  ...                                                ...   
1619  802624713319034886    https://pbs.twimg.com/media/CsrjryzWgAAZY00.jpg   
1624  803692223237865472    https://pbs.twimg.com/media/CZhn-QAWwAASQan.jpg   
1627  804413760345620481    https://pbs.twimg.com/media/CuRDF-XWcAIZSer.jpg   
1634  805958939288408065    https://pbs.twimg.com/media/CtzKC7zXEAALfSo.jpg   
1636  806242860592926720    https://pbs.twimg.com/media/Ct72q9jWcAAhlnw.jpg   
1640  807059379405148160    https://pbs.twimg.com/media/Ct2qO5PXEAE6eB0.jpg   
1645  808134635716833280    https://pbs.twimg.com/media/Cx5R8wPVEAALa9r.jpg   
1652  809808892968534016    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1683  813944609378369540    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg   
1693  816014286006976512    https://pbs.twimg.com/media/CiibOMzUYAA9Mxz.jpg   
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1712  818588835076603904    https://pbs.twimg.com/media/Crwxb5yWgAAX5P_.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1727  820446719150292993    https://pbs.twimg.com/media/CxqsX-8XUAAEvjD.jpg   
1736  821813639212650496    https://pbs.twimg.com/media/CtVAvX-WIAAcGTf.jpg   
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEA

In [36]:
# How many tweets have 1, 2, 3, 4 images ? 
image_prediction.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [37]:
# Let's see how many different values in p1
image_prediction.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
miniature_pinscher            23
Chesapeake_Bay_retriever      23
seat_belt                     22
Staffordshire_bullterrier     20
Siberian_husky                20
German_shepherd               20
web_site                      19
Cardigan                      19
teddy                         18
beagle                        18
Shetland_sheepdog             18
Maltese_dog                   18
Eskimo_dog                    18
Lakeland_terrier              17
Rottweiler                    17
Shih-Tzu                      17
kuvasz                        16
Italian_greyhound             16
          

#### Qualtiy
##### archive table:
- Missing data in columns `doggo`,`floofer`,`pupper`,`puppo` wroted as None
- Four variables in four columns `doggo`,`floofer`,`pupper`,`puppo`
- Missing data, there are 78 rows as replys and not tweets, find it via columns not null for `in_reply_to_status_id`, `in_reply_to_user_id`
- Missing data, there are 181 rows as retweets, find it via columns not null for `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`
- Missing data, there are 2297 instead of 2356 in column `expanded_urls`
- `name` column has some values like 'a', 'an', 'the', 'None'
- Duplicated `expanded_urls`

##### tweet_details table:
- There are 2354 instead of 2356 (number of records in archive table)

##### image_prediction table:
- There are 2075 instead of 2356 (number of records in archive table)
- Uppercase, Lowercase, and underscores in each words in columns p1, p2, p3
- Duplicated `jpg_url`

#### Tidiness
##### archive table:
- `tweet_id` column type is `int64` not `string`
- `timestamp` column type is `object`
- four variables in four columns `doggo`,`floofer`,`pupper`,`puppo`

##### tweet_details table:
- `retweet_count` column type is `object`
- `favorite_count` column type is `object`


##### image_prediction table:
- `tweet_id` column type is `int64` not `string`

### Clean

In [38]:
archive_copy = archive.copy()
tweet_details_copy = tweet_details.copy()
image_prediction_copy = image_prediction.copy()

#### Missing Data

**`archive`: Missing data, there are 78 rows as replys and not tweets, find it via columns not null for `in_reply_to_status_id`, `in_reply_to_user_id`**

#### Define
* Drop the 78 rows from the dataset then drop the columns `in_reply_to_status_id`, `in_reply_to_user_id`

#### Code

In [39]:
# Store the indexes in idx then use them to drop all rows
idx = archive_copy[archive_copy['in_reply_to_status_id'].notnull()].index
archive_copy.drop(idx, axis=0, inplace=True)

#### Test

In [40]:
# Should be no rows
archive_copy[archive_copy['in_reply_to_status_id'].notnull()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [41]:
# Should be no rows
archive_copy[archive_copy['in_reply_to_user_id'].notnull()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Code

In [42]:
# Then, drop the columns
archive_copy.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id'], inplace=True)

#### Test

In [43]:
# Columns should be removed
list(archive_copy)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

**`archive`: Missing data, there are 181 rows as retweets, find it via columns not null for `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`**

#### Define
* Drop the 181 rows from the dataset then drop the columns `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`

#### Code

In [44]:
# Store the indexes in idx then use them to drop all rows
idx = archive_copy[archive_copy['retweeted_status_id'].notnull()].index
archive_copy.drop(idx, axis=0, inplace=True)

#### Test

In [45]:
# Should be no rows
archive_copy[archive_copy['retweeted_status_id'].notnull()]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [46]:
# Should be no rows
archive_copy[archive_copy['retweeted_status_user_id'].notnull()]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [47]:
# Should be no rows
archive_copy[archive_copy['retweeted_status_timestamp'].notnull()]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### Code

In [48]:
# Then, drop the columns
archive_copy.drop(columns=['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True)

#### Test

In [49]:
# Columns should be removed
list(archive_copy)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

**`archive`: Missing data, there are 2297 instead of 2356 in column `expanded_urls`**

In [51]:
# Let's see the new numbers after we dropped some rows
archive_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null int64
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: int64(3), object(9)
memory usage: 213.0+ KB


In [52]:
# So, there are 3 rows having expanded_urls NaN value
# Let's display them
archive_copy[archive_copy['expanded_urls'].isnull()]

tweet_id                  timestamp  \
375   828361771580813312  2017-02-05 21:56:51 +0000   
707   785515384317313025  2016-10-10 16:20:36 +0000   
1445  696518437233913856  2016-02-08 02:18:30 +0000   

                                                 source  \
375   <a href="http://twitter.com" rel="nofollow">Tw...   
707   <a href="http://twitter.com/download/iphone" r...   
1445  <a href="http://twitter.com/download/iphone" r...   

                                                   text expanded_urls  \
375   Beebop and Doobert should start a band 12/10 w...           NaN   
707      Today, 10/10, should be National Dog Rates Day           NaN   
1445    Oh my god 10/10 for every little hot dog pupper           NaN   

      rating_numerator  rating_denominator  name doggo floofer  pupper puppo  
375                 12                  10  None  None    None    None  None  
707                 10                  10  None  None    None    None  None  
1445                10                  10  None  None    None  pupper  None

#### Define
* Go to each link manually on the browser, if tweets are there with no issues, fill these rows manually, otherwise drop them

- After checking manually, These tweets exist but they don't have images 828361771580813312 , 785515384317313025 , 696518437233913856 , So we will drop them

#### Code

In [54]:
# Store thier indexes
idx = archive_copy[archive_copy['expanded_urls'].isnull()].index
archive_copy.drop(idx, axis=0, inplace=True)

#### Test

In [56]:
# Should be no rows
archive_copy[archive_copy['expanded_urls'].isnull()]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [57]:
archive_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2094 non-null int64
timestamp             2094 non-null object
source                2094 non-null object
text                  2094 non-null object
expanded_urls         2094 non-null object
rating_numerator      2094 non-null int64
rating_denominator    2094 non-null int64
name                  2094 non-null object
doggo                 2094 non-null object
floofer               2094 non-null object
pupper                2094 non-null object
puppo                 2094 non-null object
dtypes: int64(3), object(9)
memory usage: 212.7+ KB


#### Duplicated Rows
**`archive`: Duplicated `expanded_urls`**

#### Define
* Remove duplicated expanded_urls rows

#### Code

In [66]:
# Display the duplicated rows
archive_copy[archive_copy.expanded_urls.duplicated()]

tweet_id                  timestamp  \
2212  668587383441514497  2015-11-23 00:30:28 +0000   

                                                 source  \
2212  <a href="http://vine.co" rel="nofollow">Vine -...   

                                                   text  \
2212  Never forget this vine. You will not stop watc...   

                      expanded_urls  rating_numerator  rating_denominator  \
2212  https://vine.co/v/ea0OwvPTx9l                13                  10   

     name doggo floofer pupper puppo  
2212  the  None    None   None  None

In [69]:
# Get the indexes
archive_copy.query('expanded_urls == "https://vine.co/v/ea0OwvPTx9l"')

tweet_id                  timestamp  \
657   791774931465953280  2016-10-27 22:53:48 +0000   
2212  668587383441514497  2015-11-23 00:30:28 +0000   

                                                 source  \
657   <a href="http://vine.co" rel="nofollow">Vine -...   
2212  <a href="http://vine.co" rel="nofollow">Vine -...   

                                                   text  \
657   Vine will be deeply missed. This was by far my...   
2212  Never forget this vine. You will not stop watc...   

                      expanded_urls  rating_numerator  rating_denominator  \
657   https://vine.co/v/ea0OwvPTx9l                14                  10   
2212  https://vine.co/v/ea0OwvPTx9l                13                  10   

      name doggo floofer pupper puppo  
657   None  None    None   None  None  
2212   the  None    None   None  None

In [70]:
# Remove the most recent one based on timestamp
archive_copy.drop(index=657, axis=0, inplace=True)

#### Test

In [71]:
# Should be one row
archive_copy.query('expanded_urls == "https://vine.co/v/ea0OwvPTx9l"')

tweet_id                  timestamp  \
2212  668587383441514497  2015-11-23 00:30:28 +0000   

                                                 source  \
2212  <a href="http://vine.co" rel="nofollow">Vine -...   

                                                   text  \
2212  Never forget this vine. You will not stop watc...   

                      expanded_urls  rating_numerator  rating_denominator  \
2212  https://vine.co/v/ea0OwvPTx9l                13                  10   

     name doggo floofer pupper puppo  
2212  the  None    None   None  None

In [72]:
# Should be no rows
archive_copy[archive_copy.expanded_urls.duplicated()]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

**`image_prediction`: Duplicated `jpg_url`**

#### Define
* Remove duplicated jpg_url rows

#### Code

In [76]:
# Store their indexes, then drop rows
idx = image_prediction_copy[image_prediction_copy.jpg_url.duplicated()].index
image_prediction_copy.drop(idx, axis=0, inplace=True)

#### Test

In [79]:
# Should be no rows
image_prediction_copy[image_prediction_copy.jpg_url.duplicated()]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [80]:
# Should display one record
image_prediction_copy.query('jpg_url == "https://pbs.twimg.com/ext_tw_video_thumb/675354114423808004/pu/img/qL1R_nGLqa6lmkOx.jpg"')

tweet_id                                            jpg_url  \
480  675354435921575936  https://pbs.twimg.com/ext_tw_video_thumb/67535...   

     img_num       p1   p1_conf  p1_dog                p2   p2_conf  p2_dog  \
480        1  upright  0.303415   False  golden_retriever  0.181351    True   

                   p3   p3_conf  p3_dog  
480  Brittany_spaniel  0.162084    True

In [81]:
archive_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2093 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2093 non-null int64
timestamp             2093 non-null object
source                2093 non-null object
text                  2093 non-null object
expanded_urls         2093 non-null object
rating_numerator      2093 non-null int64
rating_denominator    2093 non-null int64
name                  2093 non-null object
doggo                 2093 non-null object
floofer               2093 non-null object
pupper                2093 non-null object
puppo                 2093 non-null object
dtypes: int64(3), object(9)
memory usage: 212.6+ KB


In [85]:
tweet_details_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id          2354 non-null object
retweet_count     2354 non-null object
favorite_count    2354 non-null object
dtypes: object(3)
memory usage: 55.2+ KB


In [84]:
image_prediction_copy.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2009 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2009 non-null int64
jpg_url     2009 non-null object
img_num     2009 non-null int64
p1          2009 non-null object
p1_conf     2009 non-null float64
p1_dog      2009 non-null bool
p2          2009 non-null object
p2_conf     2009 non-null float64
p2_dog      2009 non-null bool
p3          2009 non-null object
p3_conf     2009 non-null float64
p3_dog      2009 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 162.8+ KB
